chaospy/external/openturns_.py

```diff
@@ -0,0 +1,149 @@
+    >>> correlation[0, 1] = 0.3
+    >>> copula = openturns.NormalCopula(correlation)
+    >>> dependent_distribution = openturns.ComposedDistribution(
+    ...     [openturns.Normal(), openturns.Uniform()], copula)
+    >>> chaospy.OpenTURNSDist(dependent_distribution)  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+        ...
+    chaospy...DependencyError: Stochastically dependent OpenTURNS distribution unsupported
+
+.. _OpenTURNS distribution wrapper: http://openturns.github.io/\
+openturns/latest/user_manual/_generated/openturns.ChaospyDistribution.html
+"""
+from typing import Iterable
+import numpy
+
+from ..distributions import Dist, J, DependencyError
+
+
+class openturns_dist(Dist):
+    """One dimensional OpenTURNS distribution."""
+
+    def __init__(self, distribution):
+        """
+        Args:
+            distribution (openturns.Distribution):
+                Distribution created in OpenTURNS.
+        """
+        assert distribution.isContinuous(), (
+            "Only continuous distributions are supported")
+        Dist.__init__(self)
+        self.distribution = distribution
+
+    def _pdf(self, x_loc):
+        return numpy.array(self.distribution.computePDF(
+            numpy.atleast_2d(x_loc).T).asPoint())
+
+    def _cdf(self, x_loc):
+        return numpy.array(self.distribution.computeCDF(
+            numpy.atleast_2d(x_loc).T).asPoint())
+
+    def _ppf(self, q_loc):
+        return numpy.array(self.distribution.computeQuantile(
+            q_loc[0]).asPoint())
+
+    def _bnd(self, x_loc):
+        del x_loc
+        rng = self.distribution.getRange()
+        return rng.getLowerBound()[0], rng.getUpperBound()[0]
+
+    def _mom(self, k_loc):
+        return self.distribution.getMoment(int(k_loc))[0]
+
+    def __str__(self):
+        return str(self.distribution)
+
+
+class OpenTURNSDist(J):
+    """
+    OpenTURNS distribution constructor.
+
+    Args:
+        distribution (openturns.Distribution):
+            Distribution created in OpenTURNS.
+
+    Examples:
+        >>> from openturns import ComposedDistribution, Normal
+        >>> ot_distribution = ComposedDistribution([Normal(), Normal()])
+        >>> distribution = chaospy.OpenTURNSDist(ot_distribution)
+        >>> print(distribution)  # doctest: +NORMALIZE_WHITESPACE
+        OpenTURNSDist([Normal(mu = 0, sigma = 1),
+                       Normal(mu = 0, sigma = 1)])
+    """
+
+    def __init__(self, distribution):
+        from openturns import ComposedDistribution, ContinuousDistribution
+        if isinstance(distribution, ComposedDistribution):
+            distributions = distribution.getDistributionCollection()
+            if not numpy.allclose(
+                distribution.getLinearCorrelation(),
+                numpy.eye(len(distributions))
+            ):
+                raise DependencyError("Stochastically dependent "
+                                      "OpenTURNS distribution unsupported")
```

Comment:
@regislebrun, I hope you don't mind me asking:
Is there a more correct way to check for stochastic dependencies in `openturns` than linear correlation? I am starting to look into the internals of OpenTURNS, but it is a big project and I don't have the overview I'd like.

Reply:
Hi,
When you build a multivariate distribution you can always check for dependence using the hasIndependentCopula() method. See [http://openturns.github.io/openturns/latest/user_manual/_generated/openturns.Distribution.html?highlight=distribution#openturns.Distribution.hasIndependentCopula](url)
Thanks for the initiative